Driver-only ciphers wrapup (#8680)
Mbed-TLS/mbedtls: docs/architecture/psa-migration/md-cipher-dispatch.md

* Compile-time dependencies: instead of checking `defined(MBEDTLS_PSA_CRYPTO_C)`, check `defined(MBEDTLS_PSA_CRYPTO_C) || defined(MBEDTLS_PSA_CRYPTO_CLIENT)`.
* Implementers of `MBEDTLS_PSA_CRYPTO_CLIENT` will need to provide `psa_can_do_hash()` (or a more general function `psa_can_do`) alongside `psa_crypto_init()`. Note that at this point, it will become a public interface, hence we won't be able to change it at a whim.

### Internal "block cipher" abstraction (previously known as "Cipher light")

#### Definition

> The new module is automatically enabled in `build_info.h` by modules that need

Reviewer: ```suggestion
The new module is automatically enabled in `config_adjust_legacy_crypto.h` by modules that need
```

> it, namely: CCM, GCM, only when `CIPHER_C` is not available, or the new module

Reviewer: Really minor: I find it clearer this way, but I'm not a native english speaker, so your version might be fine as well :)
```suggestion
it (namely: CCM, GCM) only when `CIPHER_C` is not available or the new module
```

an `mbedtls_cipher_id_t` in the `setup()` function, because that's how they're
identifed by callers (GCM/CCM).

#### Block cipher dual dispatch

Support for dual dispatch in the new internal module `block_cipher` is extremely similar to that in MD light.

> A block cipher context contains either a legacy module's context (AES, ARIA, Camellia) or a PSA key identifier; it has a field indicated which one is in use. All fields are private.

Reviewer: ```suggestion
A block cipher context contains either a legacy module's context (AES, ARIA, Camellia) or a PSA key identifier; it has a field indicating which one is in use. All fields are private.
```

The `engine` field is almost redundant with knowledge about `type`. However, when an algorithm is available both via a legacy module and a PSA accelerator, we will choose based on the runtime availability of the accelerator when the context is set up. This choice needs to be recorded in the context structure.

Support is determined at runtime using the new internal function
```
int psa_can_do_cipher(psa_key_type_t key_type, psa_algorithm_t cipher_alg);
```

The job of this private function is to return 1 if `hash_alg` can be performed through PSA now, and 0 otherwise. It is only defined on algorithms that are enabled via PSA. As a starting point, return 1 if PSA crypto's driver subsystem has been initialized.

> Each function in the module needs to know whether to dispatch via PSA or legacy. All function with just consult the context's `engine` field, except `setup()` which will set it according to the key type and the return value of `psa_can_do_cipher()` as discussed above.

Author: Ah, the typo was I meant to type "will just" and somehow it became "with just". I'll use your version.

Note that this assumes that an operation that has been started via PSA can be completed. This implies that `mbedtls_psa_crypto_free` must not be called while an operation using PSA is in progress.

> After calling a PSA function, call `mbedtls_cipher_error_from_psa` to convert its status code.

Reviewer: I don't think the end user has to worry about PSA error conversion:
- she/he calls functions from the block_cipher module and not PSA
- block cipher module already converts PSA errors (in case the PSA path is the one used) to legacy ones
Perhaps this sentence can be removed?
